extensions/amp-vimeo/0.1/amp-vimeo.js

```diff
@@ -13,66 +13,324 @@
+    preconnect.url('https://i.vimeocdn.com', onLayout);
     // Host that Vimeo uses to serve JS, CSS and other assets needed.
-    this.preconnect.url('https://f.vimeocdn.com', onLayout);
+    preconnect.url('https://f.vimeocdn.com', onLayout);
   }
 
-  /** @override */
+  /**
+   * @override
+   * @inheritdoc
+   */
   isLayoutSupported(layout) {
     return isLayoutSizeDefined(layout);
   }
 
+  /** @override */
+  buildCallback() {
+    installVideoManagerForDoc(this.getAmpDoc());
+  }
+
   /** @override */
   layoutCallback() {
-    const videoid = user().assert(
-        this.element.getAttribute('data-videoid'),
+    const {element} = this;
+    const vidId = user().assert(
+        element.getAttribute('data-videoid'),
         'The data-videoid attribute is required for <amp-vimeo> %s',
-        this.element);
+        element);
+
     // See
     // https://developer.vimeo.com/player/embedding
-    const iframe = this.element.ownerDocument.createElement('iframe');
-    iframe.setAttribute('frameborder', '0');
-    iframe.setAttribute('allowfullscreen', 'true');
-    iframe.src = 'https://player.vimeo.com/video/' + encodeURIComponent(
-        videoid);
+    const iframe =
+        htmlFor(element)`<iframe frameborder=0 alllowfullscreen></iframe>`;
+
+    iframe.src = `https://player.vimeo.com/video/${encodeURIComponent(vidId)}`;
+
     this.applyFillContent(iframe);
-    this.element.appendChild(iframe);
+    element.appendChild(iframe);
+
     this.iframe_ = iframe;
+    this.unlistenFrame_ = listen(this.win, 'message', this.boundOnMessage_);
+
+    this.sendCommand_('ping');
+
     return this.loadPromise(iframe);
   }
 
+  /** @override */
+  unlayoutCallback() {
+    this.removeIframe_();
+    return true; // layout again.
+  }
+
+  /** @private */
+  removeIframe_() {
+    if (this.iframe_) {
+      removeElement(this.iframe_);
+      this.iframe_ = null;
+    }
+    if (this.unlistenFrame_) {
+      this.unlistenFrame_();
+      this.unlistenFrame_ = null;
+    }
+  }
+
+  /** @private */
+  onReady_() {
+    const {element} = this;
+
+    Object.keys(VIMEO_EVENTS).forEach(event => {
+      this.sendCommand_('addEventListener', event);
+    });
+
+    Services.videoManagerForDoc(element).register(this);
+
+    element.dispatchCustomEvent(VideoEvents.LOAD);
+  }
+
+  /**
+   * @param {!Event} event
+   * @private
+   */
+  onMessage_(event) {
+    if (!isVimeoUrl(event.origin)) {
+      return;
+    }
+
+    if (event.source !== this.iframe_.contentWindow) {
+      return;
+    }
+
+    const eventData = getData(event);
+
+    if (!isJsonOrObj(eventData)) {
+      return;
+    }
+
+    const data = /** @type {?JsonObject} */ (
+      isObject(eventData) ? eventData : tryParseJson(eventData));
+
+    if (data['event'] == 'ready' || data['method'] == 'ping') {
+      this.onReady_();
+      return;
+    }
+
+    const {element} = this;
+
+    const eventToDispatch = VIMEO_EVENTS[data['event']];
+    if (eventToDispatch) {
```

Comment:
Is it worth rechecking that the event is in `VIMEO_EVENTS` or is it guaranteed they wont sent events that `addEventListener` message hasn't been sent for?

Reply:
They won't send events that you don't listen for.